### ISSUE TITLE
Switch to Keras

### ISSUE THREAD

**Issue by gillesdegottex** [gillesdegottex/percival#21]

**Comment by ljuvela** [comment]
Why not TensorFlow? I experimented with the gradient penalty for WGAN on Keras and while it worked, it was needlessly complicated. (See e.g. https://github.com/keras-team/keras-contrib/blob/master/examples/improved_wgan.py). TF lets you handle gradients and apply custom losses more freely. 

My go-to reference for WGAN-GP on TensorFlow is this one: https://github.com/tkarras/progressive_growing_of_gans
They also have their original Theano implementation available in the repo.

**Comment by gillesdegottex** [comment]
Do you mean between:
https://github.com/keras-team/keras-contrib/blob/master/examples/improved_wgan.py#L43
and
https://github.com/tkarras/progressive_growing_of_gans/blob/master/loss.py#L25
?

**Comment by gillesdegottex** [comment]
mhmm I didn't know TF has a get_output_for, I thought it was lower level like theano and Keras was quite convenient for calling a network on another one.

**Comment by ljuvela** [comment]
Wow, the gradient penalty was so much cleaner in their original Theano code https://github.com/tkarras/progressive_growing_of_gans/blob/82c0b36169413a04ad93ec6cb2af7d4260773277/train.py#L336 

You can do essentially the same in TF. Here's a snippet from my code
``` python
# Discriminator outputs
D_out_real = disc_model.forward_pass(x_real)
D_out_fake = disc_model.forward_pass(x_fake)

# Wasserstein loss
D_loss = -tf.reduce_mean(D_out_real) + tf.reduce_mean(D_out_fake) 
G_loss = -tf.reduce_mean(D_out_fake)

# Gradient penalty
epsilon = tf.random_uniform([batch_size, 1, 1], 0.001, 0.999) # different mixing factor for each sample in batch
x_hat = epsilon*x_real + (1.0-epsilon)*x_fake
d_hat = disc_model.forward_pass(x_hat)
gradients = tf.gradients(d_hat, x_hat)[0]
gradnorm = tf.sqrt(tf.reduce_sum(tf.square(gradients), axis=[1,2]))
gradient_penalty = 10*tf.reduce_mean(tf.square(gradnorm-1.0)) # magic weight factor
D_loss += gradient_penalty
```

**Comment by gillesdegottex** [comment]
oh yes, that's neater :)

**Comment by ljuvela** [comment]
The Keras example is quite neat, until it starts writing partial functions for wrapping the GP to a Keras style loss function
https://github.com/keras-team/keras-contrib/blob/b3dc815ba5c5957ed4435768030f0952dbd318e0/examples/improved_wgan.py#L238,

That's a lot of effort for working around what Keras expects loss functions to be.

**Comment by gillesdegottex** [comment]
Indeed, it seems Keras is turning around its limits here ...
Ok, I'll give a try to raw TF for simple networks first and see.
Thanks @ljuvela !!
